ipa-core/src/protocol/basics/validate.rs

```diff
@@ -0,0 +1,218 @@
+use std::{
+    convert::Infallible,
+    marker::PhantomData,
+    pin::Pin,
+    task::{Context as TaskContext, Poll},
+};
+
+use futures::{
+    future::try_join,
+    stream::{Fuse, Stream, StreamExt},
+    Future, FutureExt,
+};
+use generic_array::GenericArray;
+use pin_project::pin_project;
+use sha2::{
+    digest::{typenum::Unsigned, FixedOutput, OutputSizeUser},
+    Digest, Sha256,
+};
+
+use crate::{
+    error::Error,
+    ff::Serializable,
+    helpers::{Direction, Message},
+    protocol::{context::Context, RecordId},
+    secret_sharing::{replicated::ReplicatedSecretSharing, SharedValue},
+};
+
+type HashFunction = Sha256;
+type HashSize = <HashFunction as OutputSizeUser>::OutputSize;
+type HashOutputArray = [u8; <HashSize as Unsigned>::USIZE];
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct HashValue(GenericArray<u8, HashSize>);
+
+impl Serializable for HashValue {
+    type Size = HashSize;
+    type DeserializationError = Infallible;
+
+    fn serialize(&self, buf: &mut GenericArray<u8, Self::Size>) {
+        buf.copy_from_slice(self.0.as_slice())
+    }
+
+    fn deserialize(buf: &GenericArray<u8, Self::Size>) -> Result<Self, Self::DeserializationError> {
+        Ok(Self(buf.to_owned()))
+    }
+}
+
+impl Message for HashValue {}
+
+struct ReplicatedValidatorFinalization<C> {
+    f: Pin<Box<dyn Future<Output = Result<(), Error>>>>,
+    ctx: C,
+}
+
+impl<C: Context + 'static> ReplicatedValidatorFinalization<C> {
+    fn new(active: ReplicatedValidatorActive<C>) -> Self {
+        let ReplicatedValidatorActive {
```

Comment:
it may be nicer from API's perspective to let `ReplicatedValidatorActive` to turn itself into a pair of hashes

Reply:
As this is just internal, I didn't do that.  Though I did implement `From` for `HashValue`, which made this function a little less ugly.